In [1]:
# import modules
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Part 1: Basic Python

Question 1: Add up all the numbers from 1 to 100 (e.g. 1, 2, 3, ... , 99, 100).

In [2]:
sum_all = 0
for i in range(1, 101):
    sum_all += i
print(sum_all)

5050


Question 2: Add up all the even numbers from 1 to 100 (e.g. 2, 4, 6, ... 98, 100).

In [3]:
sum_even = 0
for i in range(2, 101):
    if i % 2 == 0:
        sum_even += i
print(sum_even)

2550


Question 3: What is the type of the variable x defined as x = 42.0?

In [4]:
x = 42.0
type(x)

float

Question 4: Consider the list my_list = [42, 47, 1.168, 3.14, 8, -5]. What code would return the first three entries of my_list?

In [5]:
my_list = [42, 47, 1.168, 3.14, 8, -5]

In [6]:
my_list[0:3]

[42, 47, 1.168]

Question 5: What code would return the last entry of my_list?

In [7]:
my_list[-1]

-5

Question 6: What is the value of $2^{8} / 23$ (rounded to two decimal places)

In [8]:
2 ** 8 / 23

11.130434782608695

In [9]:
round(2 ** 8 / 23, 2)

11.13

Question 7: Write a function that takes two arguments (x and y) and evaluates the following equation:

$\frac{1}{10^5}\sum_{n = 1}^{10^5}(\frac{y + \frac{4y}{nx}}{\frac{x}{2}})^n$

What is the value of this function when x = 4 and y = 2 (to 5 decimal places)?

In [10]:
n = 1
maxi = 10 ** 5
def calculate_sum(x,y):
    for i in range(n, maxi+1):
        sum = (1 / (10 ** 5)) * ((y + ((4 * y)/(n*x)))/(x/2)) ** n 
    return sum

In [11]:
calculate_sum(4,2)

2e-05

## Part 2: Numpy

In [12]:
np.random.seed(14) # This guarantees the code will generate the same set of random numbers whenever executed
random_integers = np.random.randint(1,high=500000, size=(100, 5))
random_integers

array([[117356,  79193, 402701, 117143, 285000],
       [306938, 482663, 340715, 187037, 489099],
       [361681, 301630, 237033, 384878, 202395],
       [455706,  66008, 310263, 272876, 121377],
       [364536, 164101, 393166, 244875, 302731],
       [ 79361,  61911, 236629, 243846, 376358],
       [271993, 313446, 268330, 493289, 426216],
       [  8772, 427728,  95557, 332750, 273582],
       [ 68988,  16146, 130812, 185335, 153473],
       [224416, 213770, 365756,  49180, 103414],
       [172251, 436012, 294647,  75763, 274690],
       [  3965, 348488, 178281, 490176, 370571],
       [100398, 417520, 305183, 447380, 386626],
       [404538, 385612, 153806, 211282, 342374],
       [ 60174, 338873, 423248, 214698, 168576],
       [476438,  43534, 254276, 165950, 274763],
       [ 65263, 386490, 268751,  57949, 343279],
       [450493, 298506,   4431, 211333, 108723],
       [270822, 481845,  16642, 329868, 165861],
       [405362, 323972, 471053, 211869,  81750],
       [313833, 1345

Question 1: What is the arithmetic mean of the second column (i.e. the column that starts with 79193) to two decimal places?

In [13]:
second_column = random_integers[:,1]
second_column

array([ 79193, 482663, 301630,  66008, 164101,  61911, 313446, 427728,
        16146, 213770, 436012, 348488, 417520, 385612, 338873,  43534,
       386490, 298506, 481845, 323972, 134546, 274024, 306390, 403450,
       152164, 306739,  83746,  45408,  13085, 417382, 191233, 471620,
        49359, 101231, 305827, 372070,  13652,  60478,  57940,  71350,
       247621, 230490, 236785, 234304,  10550,  14060, 318956, 101576,
        32365,  16347, 203736, 261539, 482596,   3921, 381271, 471600,
       104288, 131955, 170149, 322362, 323248, 214228, 402078, 283000,
       157409, 201896, 222504, 359432,  39074, 247540, 495427, 223241,
       299713, 291642,  44283, 180281, 443134, 324407,  17736, 421401,
       138691, 162707, 230751, 488671, 461726, 181060, 113532, 282380,
        98615,  55407,  12991, 473101, 323716, 263884, 254496, 457353,
       425322,   5406, 323324, 479679])

In [14]:
round(np.mean(second_column), 2)

238200.99

Question 2: What is the arithmetic mean of the first 20 rows of the third and fourth columns (the columns starting with 402701 and 117143) to two decimal places?

In [15]:
first_20 = random_integers[:20, 2:4]
first_20

array([[402701, 117143],
       [340715, 187037],
       [237033, 384878],
       [310263, 272876],
       [393166, 244875],
       [236629, 243846],
       [268330, 493289],
       [ 95557, 332750],
       [130812, 185335],
       [365756,  49180],
       [294647,  75763],
       [178281, 490176],
       [305183, 447380],
       [153806, 211282],
       [423248, 214698],
       [254276, 165950],
       [268751,  57949],
       [  4431, 211333],
       [ 16642, 329868],
       [471053, 211869]])

In [16]:
round(np.mean(first_20), 2)

251968.92

Question 3: Build a matrix with 3 columns and 8 rows with the numbers 1, 2, 3...23, 24 as entries moving across the rows. It should kinda look like this, but bigger:

$$
\begin{pmatrix}
1 &amp; 2 &amp; 3\\
4 &amp; 5 &amp; 6 \\ 
7 &amp; 8 &amp; 9 \\
\vdots &amp; \vdots &amp; \vdots \\
\end{pmatrix}
$$
What is the average (mean) of the square root of all entries the second column (to two decimal places)?

(Hint: you can make the matrix by hand, or you can combine range and reshape, which you can learn about [here](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html])!

In [17]:
ranges = (1,25)
matrix = np.arange(1,25).reshape(8,3)
matrix

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12],
       [13, 14, 15],
       [16, 17, 18],
       [19, 20, 21],
       [22, 23, 24]])

In [18]:
second_column = matrix[:, 1]
second_column

array([ 2,  5,  8, 11, 14, 17, 20, 23])

In [19]:
square_root = second_column ** 2

In [20]:
square_root

array([  4,  25,  64, 121, 196, 289, 400, 529])

In [21]:
mean = round(np.mean(square_root))

In [22]:
mean

204

## Part 3: Pandas

In [23]:
#read data
df = pd.read_csv('world-small.csv')
df.head()

country      region  gdppcap08  polityIV
0    Albania  C&E Europe       7715      17.8
1    Algeria      Africa       8033      10.0
2     Angola      Africa       5899       8.0
3  Argentina  S. America      14333      18.0
4    Armenia  C&E Europe       6070      15.0

Question 1: How many rows are in this dataset?

In [24]:
df.shape

(145, 4)

Question 2: In alphabetical order, what countries had GDP per Capita values in 2008 (the variable gdppcap08) above 17,000 and below 20,000.

Please just type the country names with the same capitalization as is found in the original data using only spaces to separate values.

In [25]:
df_gdp_btn = df[np.logical_and((df.gdppcap08 >= 17000) , (df.gdppcap08 <= 20000))]

In [26]:
df_gdp_btn.head()

country      region  gdppcap08  polityIV
31     Croatia  C&E Europe      19084      18.4
57     Hungary  C&E Europe      19330      20.0
74      Latvia  C&E Europe      17100      18.0
78   Lithuania  C&E Europe      18824      20.0
105     Poland  C&E Europe      17625      20.0

Question 3: What country in this data has the lowest GDP Per Capita?

In [27]:
df_gdp_btn.gdppcap08.min()

17100

Question 4: The polityIV variable in this data is an expert score for a country's authoritarianism. Traditionally, values of -10 are extreme autocracies, while values of 10 are consolidated liberal democracies, although in this dataset they have been rescaled to run from 0-20, where 0 is an extreme autocracy, and 20 is a consolidated liberal democracy. What is the average (rescaled) polity score for the 10 poorest countries in this data (to two decimal places)?

In [28]:
df.describe()

gdppcap08    polityIV
count    145.000000  145.000000
mean   13251.993103   13.407816
std    14802.581676    6.587626
min      188.000000    0.000000
25%     2153.000000    7.666667
50%     7271.000000   16.000000
75%    19330.000000   19.000000
max    85868.000000   20.000000

In [29]:
sort_df_gdppcap08 = df.sort_values('gdppcap08')

In [30]:
poor_10 = sort_df_gdppcap08.iloc[:10]

In [31]:
poor_10

country  region  gdppcap08   polityIV
144                  Zimbabwe  Africa        188   6.000000
29             Congo Kinshasa  Africa        321  15.000000
76                    Liberia  Africa        388  10.000000
53              Guinea-Bissau  Africa        538  11.000000
40                    Eritrea  Africa        632   3.000000
96                      Niger  Africa        684  15.333333
22   Central African Republic  Africa        736  10.200000
113              Sierra Leone  Africa        766  15.000000
129                      Togo  Africa        829   8.000000
81                     Malawi  Africa        837  15.000000

In [32]:
round(poor_10.polityIV.mean(), 2)

10.85

Question 5: What is the average polity score for the 10 richest countries (to one decimal place)?

In [33]:
rich_10 = sort_df_gdppcap08.iloc[-10:]
rich_10

country        region  gdppcap08  polityIV
6          Austria     W. Europe      38152      20.0
133            UAE   Middle East      38830       2.0
71          Kuwait   Middle East      39914       3.0
93     Netherlands     W. Europe      40849      20.0
124    Switzerland     W. Europe      42536      20.0
62         Ireland     W. Europe      44200      20.0
137  United States    N. America      46716      20.0
114      Singapore  Asia-Pacific      49284       8.0
98          Norway   Scandinavia      58138      20.0
107          Qatar   Middle East      85868       0.0

In [34]:
round(rich_10.polityIV.mean(), 1)

13.3